[qt] navigate to  transaction history page after send

Sjors opened the pull request:
Before this change QT just remained on the Send tab, which I found confusing. Now it switches to the Transactions tab. This makes it more clear to the user that the send actually succeeded, and here they can monitor progress.

Ideally I would like to highlight the transaction, e.g. by refactoring `TransactionView::focusTransaction(const QModelIndex &idx)` to accept a transaction hash, but I'm not sure how to do that.

laanwj commented:
Concept ACK

MarcoFalke commented:
utACK ddbc5aaee1bdb69583eb711ad3145fa5f8b67933

promag commented:
> just remained on the Send tab, which I found confusing

IMHO this is more confusing, having such sudden UI change. What if he wants to do multiple sends?

How about adding a checkbox like `[x] View transaction after sending` and remembering the checkbox state in settings?

promag commented on src/qt/sendcoinsdialog.h:
```diff
@@ -54,6 +54,9 @@ public Q_SLOTS:
                     const CAmount& watchOnlyBalance, const CAmount& watchUnconfBalance, const CAmount& watchImmatureBalance);
 
+Q_SIGNALS:
+    void transactionSent();
```
If the idea is to highlight the transaction, adding the txid or like can help.

Sjors replied:
I can extract the tx id and pass it along as a string. The part I couldn't figure out is how to convert that to whatever `focusTransaction(const QModelIndex &idx)` needs.

promag replied:
Nit, rename to `coinsSent()`.

Regarding row focus, see https://stackoverflow.com/a/35154534.

Sjors replied:
Done

Sjors commented:
@promag I don't think I've ever seen an online banking application that doesn't take the user to the transaction overview page after sending money.

In most financial applications Send is a modal experience, where the Send screen is pushed on top of the transaction screen and popped when the user is done. In a desktop browser that's usually a modal window, on an iPhone it's a window that animates in from the right, with an arrow on the top left to dismiss it. Doing another send is then simply a matter of clicking on Send again.

Making the experience modal might be too much of a change, but e.g. we could get rid of all tabs: merge the Overview and Transactions tab, Send and Receive open a new window.

I do think highlighting the new transaction would give a better visual clue to the user as to why they moved to this other tab.

jonasschnelli commented:
utACK ddbc5aaee1bdb69583eb711ad3145fa5f8b67933

promag commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -369,6 +369,8 @@ void SendCoinsDialog::on_sendButton_clicked()
         accept();
         CoinControlDialog::coinControl()->UnSelectAll();
         coinControlUpdateLabels();
+        // Navigate to transaction history page:
```
Remove comment? It's only relevant below when connecting.

Sjors replied:
Removed

promag reviewed (commented):
@Sjors like you say the current layout is very different from those you mention. Anyway, ACK with the row focus.

practicalswift commented:
Concept ACK

randolf commented:
Concept ACK.

Sjors commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -369,6 +369,8 @@ void SendCoinsDialog::on_sendButton_clicked()
         accept();
         CoinControlDialog::coinControl()->UnSelectAll();
         coinControlUpdateLabels();
+        // Navigate to transaction history page:
```
Coin control fields, custom change address and selected inputs, are already cleared after you send.

promag replied:
Please verify custom change address checkbox and line edit because here they are not cleared.

Sjors replied:
You're right. Turns out #12432 fixes that, because  `SendCoinsDialog::accept()` is called when the transaction is sent, which in turn calls `clear()`.

Sjors commented:
It now highlights the transaction.

promag commented on src/qt/transactionview.cpp:
```diff
@@ -592,6 +592,22 @@ void TransactionView::focusTransaction(const QModelIndex &idx)
+        TransactionTableModel::TxHashRole,
+        hash);
+
+    if (!results.length()) 
```
```cpp
if (results.isEmpty()) return;

Sjors replied:
Fixed.

promag commented on src/qt/transactionview.cpp:
```diff
@@ -592,6 +592,22 @@ void TransactionView::focusTransaction(const QModelIndex &idx)
 
+void TransactionView::focusTransaction(const QString &hash)
+{
+    if(!transactionProxyModel)
```
Space after `if`, one line.

Sjors replied:
Fixed.

promag commented on src/qt/sendcoinsdialog.h:
```diff
@@ -54,6 +54,9 @@ public Q_SLOTS:
                     const CAmount& watchOnlyBalance, const CAmount& watchUnconfBalance, const CAmount& watchImmatureBalance);
 
+Q_SIGNALS:
+    void coinsSent(QString);
```
`const QString& txid` so that it's clear what the argument is?

Sjors replied:
Done.

promag commented on src/qt/transactionview.cpp:
```diff
@@ -592,6 +592,22 @@ void TransactionView::focusTransaction(const QModelIndex &idx)
+    if(!transactionProxyModel)
+        return;
+
+    const QModelIndexList results = this->model->getTransactionTableModel()->match(
```
IIRC this performs linear search. Probably something worth to optimise in case of bigger wallets.

Sjors replied:
Are transactions in reverse chronological order? In that case it should be pretty quick, unless some race condition causes the new transaction to be missing.

luke-jr replied:
Transactions are *displayed* in whatever order the user chooses. I'm not sure what order is internally used, however...

Sjors replied:
I created a fresh wallet and then called `generate 10000`. I didn't notice any delay when sending transactions, although my machine is relatively fast. I also tried reversing transaction order in the view.

jonasschnelli commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -369,6 +370,8 @@ void SendCoinsDialog::on_sendButton_clicked()
         CoinControlDialog::coinControl()->UnSelectAll();
         coinControlUpdateLabels();
+        const uint256 txid = currentTransaction.getTransaction()->GetHash();
+        Q_EMIT coinsSent(QString::fromStdString(txid.ToString()));
```
Meh on using strings for internal uint256 transport.
Why not uint256? Could also register the type if not yet supported.

Sjors replied:
I tried passing the uint256 directly and got a bloodbath of compiler errors. How do I go about registering the type?

promag replied:
With `Q_DECLARE_METATYPE`, see https://github.com/bitcoin/bitcoin/blob/dcfe218626b05204e9fbc95ba5d95ca0eb72ec9b/src/qt/bitcoin.cpp#L80-L82

Sjors commented:
Now passing the transaction hash as `uint256`.

promag commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -13,6 +13,7 @@
 #include <qt/optionsmodel.h>
 #include <qt/platformstyle.h>
 #include <qt/sendcoinsentry.h>
+#include <qt/transactiontablemodel.h>
```
Can be removed.

promag commented on src/qt/sendcoinsdialog.h:
```diff
@@ -54,6 +54,9 @@ public Q_SLOTS:
                     const CAmount& watchOnlyBalance, const CAmount& watchUnconfBalance, const CAmount& watchImmatureBalance);
 
+Q_SIGNALS:
+    void coinsSent(const uint256 txid);
```
Use reference? `const uint256& txid`.

promag commented on src/qt/transactionview.cpp:
```diff
@@ -592,6 +591,21 @@ void TransactionView::focusTransaction(const QModelIndex &idx)
     transactionView->setFocus();
 }
 
+void TransactionView::focusTransaction(const uint256 txid)
```
Use reference? `const uint256& txid`.

promag reviewed (commented):
Tested ACK fdea382.

promag commented:
reACK b7ebe39, only changes are the suggested above.

luke-jr commented:
> What if he wants to do multiple sends?

@promag Then he should do them all at once...

luke-jr commented on src/qt/transactionview.cpp:
```diff
@@ -592,6 +591,21 @@ void TransactionView::focusTransaction(const QModelIndex &idx)
+
+    if (results.isEmpty()) return;
+
+    focusTransaction(results[0]);
```
We support selecting multiple transactions, so this ought to select all of them...

Sjors replied:
I consider more than 1 result an error, given that we're passing a single tx id. I could replace the `isEmpty()` with a check that length is 1.

luke-jr replied:
It's definitely not an error. The transaction list shows logical transactions, not blockchain transactions. There is 1 txid per blockchain transaction, but that could be many logical transactions.

Sjors replied:
Ah, I see what you mean. E.g. sending to two destinations in one transaction results in two rows in the transactions tab.

promag replied:
@luke-jr :+1:

promag commented:
@luke-jr agree, came to the same conclusion little after.

Sjors commented:
I'd like to select both entries on the transaction screen when a user sends to two addresses. Unfortunately:
* `this->model->getTransactionTableModel()->match(...` only returns one entry.  
* transactionView->selectRow(...) deselects whichever row was previously selected (even though the user can manually select multiple transactions)

luke-jr commented:
https://doc.qt.io/qt-5/qtableview.html#setSelection with `QItemSelectionModel::Rows` (and probably `QItemSelectionModel::ClearAndSelect`).

Why would match only return one entry??

Sjors commented:
> Why would match only return one entry??

I don't know. Here's a [WIP commit](https://github.com/Sjors/bitcoin/commit/8167cfd3de27da2be5852fe9561d4ae47ed361e3). The `fprintf(stderr, "row:` line is only called once.

I'm confused about the relation between `this->model->getTransactionTableModel()`, `transactionView` and `transactionProxyModel`. E.g. what does `QModelIndex targetIdx = transactionProxyModel->mapFromSource(idx);` do? Maybe @laanwj can explain the big picture here?

Sjors commented:
I added @promag's commit to enable multi row selection.

promag commented on src/qt/transactionview.cpp:
```diff
@@ -592,6 +591,25 @@ void TransactionView::focusTransaction(const QModelIndex &idx)
+
+    transactionView->setFocus();
+    transactionView->selectionModel()->clearSelection();
+    for (const QModelIndex& index : results) {
```
We could ensure first row is visible:
```cpp
for (...) {
    if (index == results[0]) transactionView->scrollTo(index);
```

Sjors replied:
That worked (using `transactionProxyModel->mapFromSource`) when transactions are ordered by date. Unfortunately when they're ordered in reverse it doesn't scroll all the way to the bottom. Maybe we need to wait a little bit for `transactionView` to become aware of its new size?

promag commented on src/qt/transactionview.cpp:
```diff
@@ -592,6 +591,25 @@ void TransactionView::focusTransaction(const QModelIndex &idx)
 
+void TransactionView::focusTransaction(const uint256& txid)
+{
+    if(!transactionProxyModel)
```
Nit, add space after `if`.

Sjors replied:
Fixed

promag commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -369,6 +369,8 @@ void SendCoinsDialog::on_sendButton_clicked()
         accept();
         CoinControlDialog::coinControl()->UnSelectAll();
         coinControlUpdateLabels();
+        const uint256 txid = currentTransaction.getTransaction()->GetHash();
```
Avoid copy here:
```cpp
const uint256& txid = ...
```

laanwj replied:
Or roll it into the Q_EMIT, as txid is not used after this. YMMV. For performance I'd be surprised it makes any difference.
```
Q_EMIT coinsSent(currentTransaction.getTransaction()->GetHash());
```

Sjors replied:
Rolled into `Q_EMIT`.

promag reviewed (commented):
Tested ACK after squash and comments addressed.

laanwj commented:
> @promag I don't think I've ever seen an online banking application that doesn't take the user to the transaction overview page after sending money.
> @promag Then he should do them all at once...

+1

laanwj reviewed (approved):
utACK

promag commented on src/qt/transactionview.cpp:
```diff
@@ -592,6 +591,27 @@ void TransactionView::focusTransaction(const QModelIndex &idx)
+        transactionView->selectionModel()->select(
+            targetIndex,
+            QItemSelectionModel::Rows | QItemSelectionModel::Select);
+        if (index == results[0]) transactionView->scrollTo(targetIndex);
```
On my system calling `transactionView->scrollTo(targetIndex)` for every model index fixes it (not sure why). It also works in the default order because scroll hint is `QAbstractItemView::EnsureVisible`, so focusing the 2nd row will not remove the 1st from the viewport (the 2nd row is already visible).

But it's also possible to order by other column (for instance address), and in that case it can be impossible to show the whole selection in the table viewport.

I see 3 possible solutions for that:
 - do nothing :trollface:;
 - automatically sort by date descending (new rows will be the first);
 - automatically fill the filter field with the new txid.

Sjors replied:
When I call `scrollTo` for every model index and transactions are sort by ascending date, things get weird: if I send to two destinations it scrolls correctly, but if I send to one destination it won't scroll far enough.

Calling `scollTo()` twice "fixes" it. Seems suspiciously similar to [this](https://forum.qt.io/topic/81671/scrollto-not-working-properly/4), I'll poke around a bit..

I prefer your first solution for the case where transactions are not ordered by date. At least one of them will be in view.

Sjors committed:
[qt] navigate to  transaction history page after send

The transaction will be selected. When sending to multiple
destinations, all will be selected (thanks @promag).
